test/integration/cni/pod_traffic_across_az_test.go

```diff
@@ -0,0 +1,305 @@
+		err = f.K8sResourceManagers.DaemonSetManager().DeleteAndWaitTillDaemonSetDeleted(testDaemonSet, utils.DefaultDeploymentReadyTimeout)
+		Expect(err).ToNot(HaveOccurred())
+
+	})
+
+	Context("While testing API Server Connectivity", func() {
+		It("Should connect to the API Server", func() {
+			// Standard paths for SA token, CA cert and API Server URL
+			token_path := "/var/run/secrets/kubernetes.io/serviceaccount/token"
+			cacert := "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+			api_server_url := "https://kubernetes.default.svc/api"
+
+			for az := range azToPod {
+				fmt.Printf("Testing API Server Connectivity from AZ %s \n", az)
+				sa_token := []string{"cat", token_path}
+				token_value, _, err := RunCommandOnPod(azToPod[az], sa_token)
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("failed to get SA token for pod in %s", az))
+				bearer := fmt.Sprintf("Authorization: Bearer %s", token_value)
+				test_api_server_connectivity := []string{"curl", "--cacert", cacert, "--header", bearer, "-X", "GET",
+					api_server_url,
+				}
+
+				api_server_stdout, _, err := RunCommandOnPod(azToPod[az], test_api_server_connectivity)
+				// Descriptive error message on failure to connect to API Server from particular AZ.
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Error while connecting to API Server from %s", az))
+				Expect(api_server_stdout).ToNot(BeEmpty())
+				Expect(api_server_stdout).To(ContainSubstring("APIVersions"))
+			}
+		})
+	})
+})
+
+func CheckConnectivityBetweenPods(azToPod map[string]coreV1.Pod, port int, testerExpectedStdOut string, testerExpectedStdErr string, getTestCommandFunc func(serverPod coreV1.Pod, port int) []string) {
+
+	By("checking connection on same node, primary to primary")
+
+	for az1 := range azToPod {
+		for az2 := range azToPod {
+			if az1 != az2 {
+				fmt.Printf("Testing Connectivity from Pod IP1 %s (%s) to Pod IP2 %s (%s) \n",
+					azToPod[az1].Status.PodIP, az1, azToPod[az2].Status.PodIP, az2)
+				testConnectivity(azToPod[az1], azToPod[az2], testerExpectedStdOut, testerExpectedStdErr, port, getTestCommandFunc)
+			}
+		}
+	}
+}
+
+func RunCommandOnPod(receiverPod coreV1.Pod, command []string) (string, string, error) {
+	stdout, stderr, err := f.K8sResourceManagers.PodManager().
+		PodExec(receiverPod.Namespace, receiverPod.Name, command)
+	return stdout, stderr, err
+}
+
+func VerifyConnectivityForNegativeCase(azToPod map[string]coreV1.Pod, port int,
+	getTestCommandFunc func(receiverPod coreV1.Pod, port int) []string) {
+
+	for az1 := range azToPod {
+		for az2 := range azToPod {
+			if az1 != az2 {
+				fmt.Printf("Testing Connectivity from Pod IP1 %s (%s) to Pod IP2 %s (%s) \n",
+					azToPod[az1].Status.PodIP, az1, azToPod[az2].Status.PodIP, az2)
+
+				senderPod := azToPod[az1]
+				receiverPod := azToPod[az2]
+				testerCommand := getTestCommandFunc(receiverPod, port)
+
+				GinkgoWriter.Printf("verifying connectivity fails from pod %s on node %s with IP %s to pod"+
```

Comment:
Why is this case failing? Isn't this just cross-AZ pod traffic?

Reply:
Same Q, but I'm not following what this test is trying to do..and why will this fail?